autogpt/llm/providers/openai.py

```diff
@@ -348,5 +351,74 @@ def get_openai_command_specs(agent: Agent) -> list[OpenAIFunctionSpec]:
                 for param in command.parameters
             },
         )
-        for command in agent.command_registry.commands.values()
+        for command in command_registry.commands.values()
     ]
+
+
+def count_openai_functions_tokens(
+    functions: list[OpenAIFunctionSpec], for_model: str
+) -> int:
+    """Returns the number of tokens taken up by a set of function definitions
+
+    Reference: https://community.openai.com/t/how-to-calculate-the-tokens-when-using-function-call/266573/18
+    """
+    from autogpt.llm.utils import count_string_tokens
+
+    return count_string_tokens(
+        f"# Tools\n\n## functions\n\n{format_function_specs_as_typescript_ns(functions)}",
+        for_model,
+    )
+
+
+def format_function_specs_as_typescript_ns(functions: list[OpenAIFunctionSpec]) -> str:
```

Comment:
This probably belongs as a property on the `OpenAIFunctionSpec` as something like `OpenAIFunctionSpec.prompt_format` (i.e. it may not stay a typescript implementation and we don't necessarily care that is its).  

If we do that, then, we don't need a special token counting function for openai (you can delete  `count_openai_functions_tokens`) and in `chat_with_ai` this

```python
    # Account for tokens used by OpenAI functions
    openai_functions = get_openai_command_specs(agent.command_registry, agent.config)
    if openai_functions:
        functions_tlength = count_openai_functions_tokens(openai_functions, model)
        current_tokens_used += functions_tlength
        logger.debug(f"OpenAI Functions take up {functions_tlength} tokens in API call")
```

becomes

```python
    # Account for tokens used by OpenAI functions
    openai_functions = get_openai_command_specs(agent.command_registry, agent.config)
    functions_tlength = sum([count_string_tokens(f.prompt_format()) for f in openai_functions])
    logger.debug(f"OpenAI Functions take up {functions_tlength} tokens in API call")
```

Reply:
This is an optional change.

Reply:
It's not that simple. The individual function definitions go in a wrapper that also takes up N tokens.

I'll move the formatting for the individual functions into the `OpenAIFunctionSpec` object, and leave the rest as is for now.